dali/kernels/signal/moving_mean_square_gpu.cu

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda_runtime.h>
+#include <cstdio>
+#include <iostream>
+#include <numeric>
+#include <vector>
+#include "dali/core/convert.h"
+#include "dali/core/util.h"
+#include "dali/kernels/signal/moving_mean_square.h"
+#include "dali/kernels/signal/moving_mean_square_gpu.h"
+
+#define SHARED_MEMORY_BANKS 32
+#define LOG_MEM_BANKS 5
+#define CONFLICT_FREE_OFFSET(n) ((n) >> LOG_MEM_BANKS)
+
+namespace dali {
+namespace kernels {
+namespace signal {
+
+/**
+ * @brief Sample descriptor
+ */
+struct SampleDesc {
+  void *out;
+  const void *in;
+  int64_t len;
+};
+
+/**
+ * @brief Shared memory access pattern to avoid bank conflicts.
+ * @remarks See Example 39-4 from
+ *  https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-39-parallel-prefix-sum-scan-cuda
+ */
+struct conflict_free_pos {
+  DALI_HOST_DEV DALI_FORCEINLINE int operator()(int pos) const noexcept {
+    return pos + CONFLICT_FREE_OFFSET(pos);
+  }
+};
+
+struct square {
+  template <typename T>
+  DALI_HOST_DEV DALI_FORCEINLINE T operator()(T x) const noexcept {
+    return x * x;
+  }
+};
+
+struct divide {
+  explicit divide(float divisor) {
+    factor_ = 1.0 / divisor;
+  }
```

Comment:
1. I think this should be trivially default-constructible to make it less problematic in host<->device interactions.
2. There's no reason the constructor should be host-only
3. True division by `float` done in `double` and rounded will not be more precise than a true division in float, just more expensive.
```suggestion
  divide() = default;
  constexpr DALI_HOST_DEV explicit divide(float divisor) {
    factor_ = 1.0f / divisor;
  }
```